pkg/nsx/services/subnet/subnet.go

```diff
@@ -0,0 +1,243 @@
+package subnet
+
+import (
+	"errors"
+	"os"
+	"sync"
+
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt/model"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
+	"github.com/vmware-tanzu/nsx-operator/pkg/logger"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/realizestate"
+	"github.com/vmware-tanzu/nsx-operator/pkg/util"
+)
+
+var (
+	log                       = logger.Log
+	MarkedForDelete           = true
+	EnforceRevisionCheckParam = false
+	ResourceTypeSubnet        = common.ResourceTypeSubnet
+	NewConverter              = common.NewConverter
+	// Default static ip-pool under Subnet.
+	ipPoolID        = "static-ipv4-default"
+	SubnetTypeError = errors.New("unsupported type")
+)
+
+type SubnetService struct {
+	common.Service
+	SubnetStore *SubnetStore
+}
+
+// SubnetParameters stores parameters to CRUD Subnet object
+type SubnetParameters struct {
+	OrgID     string
+	ProjectID string
+	VPCID     string
+}
+
+var subnetService *SubnetService
+var lock = &sync.Mutex{}
+
+// GetSubnetService get singleton SubnetService instance, subnet/subnetset controller share the same instance.
+func GetSubnetService(service common.Service) *SubnetService {
+	if subnetService == nil {
+		lock.Lock()
+		defer lock.Unlock()
+		if subnetService == nil {
+			var err error
+			if subnetService, err = InitializeSubnetService(service); err != nil {
+				log.Error(err, "failed to initialize subnet commonService")
+				os.Exit(1)
+			}
+		}
+	}
+	return subnetService
+}
+
+// InitializeSubnetService initialize Subnet service.
+func InitializeSubnetService(service common.Service) (*SubnetService, error) {
+	wg := sync.WaitGroup{}
+	wgDone := make(chan bool)
+	fatalErrors := make(chan error)
+	subnetService := &SubnetService{
+		Service: service,
+		SubnetStore: &SubnetStore{
+			ResourceStore: common.ResourceStore{
+				Indexer: cache.NewIndexer(keyFunc, cache.Indexers{
+					common.TagScopeSubnetCRUID: subnetIndexFunc,
+				}),
+				BindingType: model.VpcSubnetBindingType(),
+			},
+		},
+	}
+
+	wg.Add(1)
+	go subnetService.InitializeResourceStore(&wg, fatalErrors, ResourceTypeSubnet, nil, subnetService.SubnetStore)
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+	select {
+	case <-wgDone:
+		break
+	case err := <-fatalErrors:
+		close(fatalErrors)
+		return subnetService, err
+	}
+	return subnetService, nil
+}
+
+// TODO Test update of VpcSubnet(eg. update tags)
+func (service *SubnetService) CreateOrUpdateSubnet(obj client.Object, vpcInfo *common.VPCResourceInfo, tags []model.Tag) (string, error) {
+	value, ok := obj.GetAnnotations()[common.TagScopeSubnetType]
+	if ok {
+		tags = append(tags, model.Tag{
+			Scope: String(common.TagScopeSubnetType),
```

Comment:
Why we need add subnet type in tag?

Reply:
Not needed any more, use label instead according to today's discusssion.